### PR TITLE
Merge branch 'master' of https://github.com/mmerickel/wired into docs_samples

### PR DESCRIPTION
**Pull request by pauleveritt**
The start of some docs re-org. Next step is to take my non-dataclass examples 
and start a "tutorial" section.

### CODE REVIEW

**Comment by pauleveritt** [comment]
@mmerickel I plan to do the tutorial RST without inline code, instead pointing to src/wired/samples 
so that I can keep the minimal tests on the samples. Perhaps we adjust the setup.py to keep those 
out of the release?

**Comment by pauleveritt** [comment]
I just pushed a completion of this. Each sample from before (minus the declarative dataclass DI stuff) is now a separate tutorial page, pointing at code, with a simple test.

**Comment by mmerickel** [comment]
With respect to the RST line length, Pyramid changed to one sentence per line. I'd be fine with that but the 80-char line breaks on RST are just a pain in the ass to maintain.

**Comment by mmerickel** [comment]
I'm fine with placing the samples in the src if you can't find another way to do it but yes I do not want them included in the wheels. What's the reason to not put them in the docs folder?

**Comment by pauleveritt** [comment]
Your project, you set the rules. :) I'll move everything back to one line and change my docs to one line.

**Comment by mmerickel** [comment]
> Your project, you set the rules. :) I'll move everything back to one line and change my docs to one line.

Well I was compromising in the middle anyway. The lines you changed were paragraph-per-line. :-)

**Comment by pauleveritt** [comment]
> I'm fine with placing the samples in the src if you can't find another way to do it but yes I do not want them included in the wheels. What's the reason to not put them in the docs folder?

Writing tests that ensure the sample works.

**Comment by pauleveritt** [comment]
> > Your project, you set the rules. :) I'll move everything back to one line and change my docs to one line.
> 
> Well I was compromising in the middle anyway. The lines you changed were paragraph-per-line. :-)

Ah, got it. I'll move to that.

**Comment by mmerickel** [comment]
For sure, so if we put them in the docs folder what do we have to change in the tests? Does it just make them harder to import?

**Comment by pauleveritt** [comment]
> For sure, so if we put them in the docs folder what do we have to change in the tests? Does it just make them harder to import?

Yeh, I don't really know how we would do an import. Open to suggestions. I guess some sys.path.insert() trick. Maybe pytest has a plugin that helps.

**Comment by mmerickel** [comment]
Yeah it's a conundrum but I think if it's just import issues then we can fix that with import tricks in those specific test files maybe in a fixture or just globally in a `conftest.py`. Either way I think the tests should skip if they can't find the code so that the tests can be run against the wheel which wouldn't include the sample code that the tests require.

**Comment by pauleveritt** [comment]
I pushed a commit with one-sentence per line.

I pushed another commit moving the samples into docs/tutorial and adjusting the test to add it to 
the path. BUT...when tox runs the tests on the sdist it won't have the docs. So those tests fail.

We need a marker strategy so that building the wheel doesn't run those tests. But first question: do you want the CI to ensure the tutorials aren't broken? If you don't care, then we'll mark them as "never run" and I'll manually run them here.

I also need to shut up the dataclass failures on <3.7.

File: CHANGES.rst [hunk 1]
```diff
@@ -1,3 +1,10 @@
+Unreleased
+==========
+
+- Edit docs to (a) improve sales pitch, (b) split into a couple of sub-pages,
+  and (c) provide a tutorial. Update README and ``setuup.py`` description a
```
Comment (mmerickel) [inline CHANGES.rst]: `setuuuuuuuuuuuuuuuuuuuuuuuuuup.py`

File: bs-config.json [hunk 2]
```diff
@@ -0,0 +1,6 @@
+{
+  "server": {
+    "baseDir": "docs/_build"
+  },
+  "open": false
+}
```
Comment (mmerickel) [inline bs-config.json]: what's this? should it be in gitignore?

**Comment by pauleveritt** [comment]
You're right, I doubt anybody but me will want it. I use it to do live reloading of my sphinx pages as I author. I'll remove it.

**Comment by pauleveritt** [comment]
Fixed and fixed (browsersync, setuuuuuuuup.)

**Comment by mmerickel** [comment]
Ah well I fixed the import path but I guess we need to skip those tests on < py37. There are some good patterns for doing that in pytest.

**Comment by pauleveritt** [comment]
Hmm, skipIf and Python versions works fine, except variable annotations blows up even parsing on older Pythons. We need tox to skip tests/tutorial if the version is older than 3.6. I'll read up on tox.ini to see how to pass command-line flags to exclude dirs on a per-version basis. (Or maybe you know.)

**Comment by mmerickel** [comment]
I'd just use https://docs.pytest.org/en/latest/skipping.html#skip-all-test-functions-of-a-class-or-module and put `pytestmark = skipif...` at module-level in `conftest.py` of the tutorial tests.

**Comment by mmerickel** [comment]
If that still breaks because it's importing the files then yeah we'll need a way to tell it conditionally to skip the tests on certain versions via the pytest config in `setup.cfg` or via conditional args to `pytest` in `tox.ini`.

**Comment by mmerickel** [comment]
I found the solution here https://docs.pytest.org/en/latest/skipping.html#skipping-files-or-directories

**Comment by pauleveritt** [comment]
Thanks, I'm struggling to get pyenv producing all the versions locally to use in tox. I suppose I should hook my fork up to Travis to avoid spamming you so much.

**Comment by mmerickel** [comment]
I use https://github.com/mmerickel/dotfiles/blob/master/aliasrc#L28-L38, I know bert has something similar that works for bash instead of zsh.

**Comment by pauleveritt** [comment]
For the lint failure on 3.6, how do you feel about me putting the imports in the fixtures and tests? Too pedantic? Can I put them at the top of the file and be a little less clean?

**Comment by mmerickel** [comment]
I don't mind if you put them at the top of the file as far as a pattern, but if you just want to fix linting in general you just need to run `tox -e black`.

**Comment by mmerickel** [comment]
Is this ready to merge then?

**Comment by pauleveritt** [comment]
Yes!! Finally, please merge. Sorry for all the crap. It started with a bad PR title and went downhill from there. :)

**Comment by mmerickel** [comment]
Lol, thanks so much for the work @pauleveritt the end result is awesome.

**Pull request merged**